modules/3d/perf/perf_rendering.cpp

```diff
@@ -0,0 +1,215 @@
+                                                       };
+
+    static inline void PrintTo(const OutputsEnum &t, std::ostream *os) { t.PrintTo(os); }
+}
+
+
+// resolution, shading type, outputs needed
+typedef perf::TestBaseWithParam<std::tuple<std::tuple<int, int>, ShadingTypeEnum, OutputsEnum>> RenderingTest;
+
+PERF_TEST_P(RenderingTest, rasterizeTriangles, ::testing::Combine(
+    ::testing::Values(std::make_tuple(1920, 1080), std::make_tuple(1024, 768), std::make_tuple(640, 480)),
+    ShadingTypeEnum::all(),
+    OutputsEnum::all()
+    ))
+{
+    auto t = GetParam();
+    auto wh = std::get<0>(t);
+    int width = std::get<0>(wh);
+    int height = std::get<1>(wh);
+    auto shadingType = std::get<1>(t);
+    auto outputs = std::get<2>(t);
+
+    string objectPath = findDataFile("rendering/spot.obj");
+
+    Vec3f position = Vec3f(20.0, 60.0, -40.0);
+    Vec3f lookat   = Vec3f( 0.0,  0.0,   0.0);
+    Vec3f upVector = Vec3f( 0.0,  1.0,   0.0);
+
+    double fovy = 45.0;
+
+    std::vector<Vec3f> vertices;
+    std::vector<Vec3i> indices;
+    std::vector<Vec3f> colors;
+    std::vector<vector<int>> indvec;
+    loadMesh(objectPath, vertices, colors, indvec);
+    for (const auto &vec : indvec)
+    {
+        indices.push_back({vec[0], vec[1], vec[2]});
+    }
+
+    if (shadingType == ShadingType::White)
+    {
+        colors.clear();
+    }
+    else
+    {
+        for (auto &color : colors)
+        {
+            color = Vec3f(abs(color[0]), abs(color[1]), abs(color[2])) * 255.0f;
+        }
+    }
+
+    double zNear = 0.1, zFar = 50;
+
+    Mat cameraMatrix = Mat(makeCamMatrix_TODO_rewrite_it_later(position, lookat, upVector, fovy, zNear, zFar));
+
+    Mat depth_buf, color_buf;
+    while (next())
+    {
+        // Prefilled to measure pure rendering time w/o allocation and clear
+        depth_buf = Mat(height, width, CV_32F, zFar);
+        color_buf = Mat(height, width, CV_32FC3, Scalar(0.0, 0.0, 0.0));
+
+        startTimer();
+        cv::triangleRasterize(vertices, indices, colors, cameraMatrix, width, height,
+                              (shadingType == ShadingType::Shaded),
+                              (outputs != Outputs::ColorOnly) ? depth_buf : noArray(),
+                              (outputs != Outputs::DepthOnly) ? color_buf : noArray());
+        stopTimer();
+    }
+
+    if (debugLevel > 0)
+    {
+        color_buf.convertTo(color_buf, CV_8UC3, 1.0f);
+        cvtColor(color_buf, color_buf, cv::COLOR_RGB2BGR);
+        cv::flip(color_buf, color_buf, 0);
+        depth_buf.convertTo(depth_buf, CV_8UC1, 1.0);
+        cv::flip(depth_buf, depth_buf, 0);
+
+        std::string shadingName;
+        {
+            std::stringstream ss;
+            shadingType.PrintTo(&ss);
+            ss >> shadingName;
+        }
+        std::string widthStr  = std::to_string(width);
+        std::string heightStr = std::to_string(height);
+        std::string suffix = widthStr + "x" + heightStr + "_" + shadingName;
```

Comment:
I recommend to use cv::format for file name generation.

Reply:
Makes sense, will fix this